File: prep/implementation/patterns_reference_implementation/in_place_reversal_of_linked_list/reverse_linked_list.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

struct ListNode {
  int val;
  ListNode *next;

  ListNode(int val) : val(val), next(nullptr) {}
};

ListNode *reverse(ListNode *head) {
  if (head == nullptr || head->next == nullptr) {
    return head;
  }

  ListNode *prev = nullptr, *curr = head, *next = nullptr;
  while (curr != nullptr) {
    next = curr->next;
    curr->next = prev;
    prev = curr;
    curr = next;
  }

  return prev;
}

void display(ListNode *head) {
  while (head != nullptr) {
    cout << head->val << " ";
    head = head->next;
  }
  cout << "\n";
}

// O(n) / O(1)
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, val;
  cin >> tt;

  while (tt--) {
    cin >> n;

    cin >> val;
    ListNode *head = new ListNode(val);
    ListNode *curr = head;

    for (int i = 1; i < n; i++) {
      cin >> val;
      curr->next = new ListNode(val);
      curr = curr->next;
    }

    head = reverse(head);
    display(head);
  }

  return 0;
}

File: lc/coreprep/implementation/practice_iterations/patterns_iterations_2/ReverseLinkedList.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class ReverseLinkedList {
  static class ListNode {
    int val;
    ListNode next;

    ListNode(int val) {
      this.val = val;
      this.next = null;
    }
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();

        ListNode head = null, tail = null;
        for (int i = 0; i < n; i++) {
          int val = in.nextInt();

          if (head == null) {
            head = tail = new ListNode(val);
          } else {
            tail.next = new ListNode(val);
            tail = tail.next;
          }
        }

        ListNode reversedHead = reverse(head);
        display(reversedHead);
      }
    }
  }

  // O(n) / O(1)
  public static ListNode reverse(ListNode head) {
    ListNode prev = null, curr = head, next = null;
    while (curr != null) {
      next = curr.next;
      curr.next = prev;
      prev = curr;
      curr = next;
    }

    return prev;
  }

  private static void display(ListNode head) {
    while (head != null) {
      System.out.printf("%d ", head.val);
      head = head.next;
    }
    System.out.println();
  }
}